compiler/z/codegen/S390Peephole.cpp

```diff
@@ -1349,29 +1349,94 @@ TR_S390PostRAPeephole::ConditionalBranchReduction(TR::InstOpCode::Mnemonic branc
+   bool branchTakenPerformReduction = false;
+   bool fallThroughPerformReduction = false;
+
+   if (_cursor->getOpCodeValue() == TR::InstOpCode::CLR
+       && _cursor->getNext()->getOpCodeValue() == TR::InstOpCode::BRC)
       {
-      printf("Finding CR + BRC\n");
-      printf("method=%s\n", comp()->signature());
-      }
+      TR::Instruction *clrInstruction = _cursor;
+      TR::Instruction *brcInstruction = _cursor->getNext();
+      TR::LabelSymbol *labelSymbol = brcInstruction->getLabelSymbol();
+      /* if branch target is a snippet we'll go ahead with the
+       * reduction. Otherwise, we look at next 5 (or until TR::BBEnd
+       * depending on whichever comes first) to check if any
+       * instruction reads or sets the condition code.
+       * We'll perform the reduction if an instruction sets
+       * the condition code or, if none of the instructions read the
+       * condition code set by clrInstruction
+       */
+      if (labelSymbol->getSnippet())
+         {
+         branchTakenPerformReduction = true;
+         }
+      else
+         {
+         // check branch taken case for condition code usage
+         TR::Instruction* branchInstruction = labelSymbol->getInstruction();
+         for (auto branchTakenInstIndex = 0; branchTakenInstIndex < 5 && NULL != branchInstruction; ++branchTakenInstIndex)
+            {
+            if (branchInstruction->getOpCode().readsCC())
+               {
+               break;
+               }
+            // CC is set before it is read (ordering of the if checks matter)
+            if (branchInstruction->getOpCode().setsCC() || TR::BBEnd == branchInstruction->getNode()->getOpCodeValue())
+               {
+               branchTakenPerformReduction = true;
+               break;
+               }
+            branchInstruction = branchInstruction->getNext();
+            }
+         }
+      // check fall through case for condition code usage
+      TR::Instruction* fallThroughInstruction = brcInstruction->getNext();
+      for (auto fallThroughInstIndex = 0; fallThroughInstIndex < 5 && NULL != fallThroughInstruction; ++fallThroughInstIndex)
+         {
+         if (fallThroughInstruction->getOpCode().readsCC())
+            {
+            break;
+            }
+         // CC is set before it is read (ordering of the if checks matter)
+         if (fallThroughInstruction->getOpCode().setsCC() || TR::BBEnd == fallThroughInstruction->getNode()->getOpCodeValue())
+            {
+            fallThroughPerformReduction = true;
+            break;
+            }
+         fallThroughInstruction = fallThroughInstruction->getNext();
+         }
 
-   return performed;
+      if (fallThroughPerformReduction && branchTakenPerformReduction)
+         {
+         TR_ASSERT_FATAL(clrInstruction->getNumRegisterOperands() == 2, "Number of register operands was not 2: %d\n", clrInstruction->getNumRegisterOperands());
+
+         TR::Instruction *clrjInstruction = generateRIEInstruction(
```

Comment:
Sorry, we are almost there. Totally forgot to mention this but could we guard the transformation with `performTransformation` here? This will help with debugging if there are any issues. Let me know if you are unfamiliar with this API and I can explain. Or even better, submit a PR with some documentation as I can see we currently do not have any.